### PR TITLE
Generalized MOOSE Plugin and Related Examples

### PR DESCRIPTION
**Pull request by miaoyinb**
This PR involves a series updates on MOOSE plugin:

- A generalized MOOSE plugin is added to replace the original SAM specified plugin;
- A simple BISON example is added in addition to the SAM example to demo the new generalized MOOSE plugin;
- Add option to enable multi-processor running of MOOSE apps thru `mpiexec -n n_cpu`;
- Add option to specify supplementary input files (e.g., exodus mesh file, sub-application input file, XS file, etc.) for MOOSE plugin;
- Add a simple MOOSE MultiApps example.

### CODE REVIEW

**Comment by miaoyinb** [comment]
@nstauff

**Comment by miaoyinb** [comment]
Note that the `#12` references in the commit messages were for previous Gitlab repo's issues system and are thus obsolete here.

**Comment by nstauff** [comment]
@miaoyinb thanks - this looks great to me. One request would be to update the [__documentation__](https://github.com/watts-dev/watts/blob/development/doc/source/user/usage.rst) and change `SAM Plugin` to `MOOSE Plugin`.
@paulromano @lingzou @zhieejhia93  - can you guys give a quick review of this MR?

**Comment by nstauff** [comment]
I approve this MR.

**Review by nstauff — APPROVED** [review]

**Review by zhieejhia93 — APPROVED** [review]

File: src/watts/plugin_moose.py [hunk 1]
```diff
@@ -106,54 +106,60 @@ def _from_hdf5(cls, obj: h5py.Group):
...
     ----------
-    sam_exec
-        Path to SAM executable
+    moose_exec
+        Path to MOOSE executable
 
     """
     def  __init__(self, template_file: str, show_stdout: bool = False,
-                  show_stderr: bool = False):
+                  show_stderr: bool = False, n_cpu: int = 1, supp_inputs: List[str] = []):
```
Comment (paulromano) [inline src/watts/plugin_moose.py:132]: Can you add a description of these parameters in the docstring above?

File: src/watts/plugin_moose.py [hunk 2]
```diff
@@ -106,54 +106,60 @@ def _from_hdf5(cls, obj: h5py.Group):
...
         super().__init__(template_file)
-        self._sam_exec = Path('sam-opt')
-        self.sam_inp_name = "SAM.i"
+        self._moose_exec = Path('moose-opt')
+        self.moose_inp_name = "MOOSE.i"
         self.show_stdout = show_stdout
         self.show_stderr = show_stderr
+        if n_cpu < 1:
+            raise RuntimeError(f"The CPU number used to run MOOSE app must be a natural number.")
```
Comment (paulromano) [inline src/watts/plugin_moose.py]:
```suggestion
            raise RuntimeError("The CPU number used to run MOOSE app must be a natural number.")
```

File: src/watts/plugin_moose.py [hunk 3]
```diff
@@ -106,54 +106,60 @@ def _from_hdf5(cls, obj: h5py.Group):
...
+        self.n_cpu = n_cpu
+        self.supp_inputs = []
+        for sif in supp_inputs:
+            self.supp_inputs = self.supp_inputs + [Path(os.path.abspath(sif))]
```
Comment (paulromano) [inline src/watts/plugin_moose.py]:
```suggestion
        self.supp_inputs = [Path(f).resolve() for f in supp_inputs]
```

File: src/watts/plugin_moose.py [hunk 4]
```diff
@@ -163,38 +169,37 @@ def prerun(self, params: Parameters):
...
 
-        log_file = Path("SAM_log.txt")
+        log_file = Path("MOOSE_log.txt")
 
         with log_file.open("w") as outfile:
             func_stdout = tee_stdout if self.show_stdout else redirect_stdout
             func_stderr = tee_stderr if self.show_stderr else redirect_stderr
             with func_stdout(outfile), func_stderr(outfile):
-                run_proc([self.sam_exec, "-i", self.sam_inp_name])
+                run_proc(["mpiexec", "-n", str(self.n_cpu) , self.moose_exec, "-i", self.moose_inp_name])
```
Comment (paulromano) [inline src/watts/plugin_moose.py:187]: Is it safe to assume that MOOSE will always be run with MPI? i.e., is there ever a case where someone doesn't want to use mpiexec?

File: src/watts/results.py [hunk 5]
```diff
@@ -1,5 +1,5 @@
-# SPDX-FileCopyrightText: 2022 UChicago Argonne, LLC
-# SPDX-License-Identifier: MIT
+#SPDX - FileCopyrightText : 2022 UChicago Argonne, LLC
+#SPDX - License - Identifier : MIT
```
Comment (paulromano) [inline src/watts/results.py]:
```suggestion
#SPDX-FileCopyrightText: 2022 UChicago Argonne, LLC
#SPDX-License-Identifier: MIT
```

File: src/watts/results.py [hunk 6]
```diff
@@ -50,9 +50,9 @@ def move_files(self, dst: PathLike):
         """
 
         dst_path = Path(dst)
-        # Move input/output files and change base -- note that trying to use the
-        # Path.replace method doesn't work across filesystems, so instead we use
-        # shutil.move
+#Move input / output files and change base-- note that trying to use the
+#Path.replace method doesn't work across filesystems, so instead we use
+#shutil.move
```
Comment (paulromano) [inline src/watts/results.py]: Not sure what happened here -- can you move this back to where it was before?

**Review by paulromano — CHANGES_REQUESTED** [review]
Thanks for the update @miaoyinb! A few small requests:

File: src/watts/plugin_moose.py (hunk 1 above)
Comment (miaoyinb) [reply src/watts/plugin_moose.py:132]: will do.

File: src/watts/results.py (hunk 6 above)
Comment (miaoyinb) [reply src/watts/results.py]: ok

File: src/watts/plugin_moose.py (hunk 4 above)
Comment (miaoyinb) [reply src/watts/plugin_moose.py:187]: MOOSE requires MPI wrapper compilers during installation, so it should be safe using this approach.

**Comment by miaoyinb** [comment]
@paulromano Please take a lot and let me know if any additional revision is needed. Thanks in advance.

**Review by paulromano — APPROVED** [review]
Thanks @miaoyinb!

**Pull request merged**